agent/csi/volumes.go

```diff
@@ -0,0 +1,248 @@
+	defer r.mu.Unlock()
+	ctx := context.Background()
+	for _, volume := range volumes {
+		v := volume.Copy()
+		log.G(ctx).WithField("method", "(*volumes).Add").Debugf("Add Volume:%v", volume.VolumeID)
+
+		r.m[volume.VolumeID] = v
+		driverName := v.Driver.Name
+		if _, ok := r.pluginMap[driverName]; !ok {
+			// TODO - revisit NodePlugin constructor call upon deciding where it needs to initialized.
+			// On deciding, will use NodeGetInfo() on plugin object accordingly to get NodeInfo and susbequently, fetch nodeID.
+			// Until then, use volume ID as node ID in lazy initialzation.
+			r.pluginMap[driverName] = NewNodePlugin(driverName, v.ID)
+		}
+		volumeObjects = append(volumeObjects, v)
+	}
+	r.iterateVolumes(volumeObjects, true)
+}
+
+func (r *volumes) iterateVolumes(volumeObjects []*api.VolumeAssignment, isAdd bool) {
+	ctx := context.Background()
+	for _, v := range volumeObjects {
+		if isAdd {
+			go r.tryAddVolume(ctx, v)
+		} else {
+			go r.tryRemoveVolume(ctx, v)
+		}
+	}
+}
+
+func (r *volumes) tryAddVolume(ctx context.Context, assignment *api.VolumeAssignment) {
+	r.tryMu.Lock()
+	defer r.tryMu.Unlock()
+
+	r.mu.RLock()
+	plugin, ok := r.pluginMap[assignment.VolumeID]
+	r.mu.RUnlock()
+	if !ok {
+		log.G(ctx).Debugf("plugin not found for VolumeID:%v", assignment.VolumeID)
+		return
+	}
+	if err := plugin.NodeStageVolume(ctx, assignment); err != nil {
+		waitFor := initialBackoff
+	retryStage:
+		for i := 0; i < maxRetries; i++ {
+			select {
+			case <-ctx.Done():
+				// selecting on ctx.Done() allows us to bail out of retrying early
+				return
+			case <-time.After(waitFor):
+				// time.After is better than using time.Sleep, because it blocks
+				// on a channel read, rather than suspending the whole
+				// goroutine. That lets us do the above check on ctx.Done().
+				//
+				// time.After is convenient, but it has a key problem: the timer
+				// is not garbage collected until the channel fires. this
+				// shouldn't be a problem, unless the context is canceled, there
+				// is a very long timer, and there are a lot of other goroutines
+				// in the same situation.
+				if err := plugin.NodeStageVolume(ctx, assignment); err == nil {
+					break retryStage
+				}
+			}
+			// if the exponential factor is 2, you can avoid using floats by
+			// doing bit shifts. each shift left increases the number by a power
+			// of 2. we can do this because Duration is ultimately int64.
+			waitFor = waitFor << 1
+		}
+	}
+
+	// Publish
+	if err := plugin.NodePublishVolume(ctx, assignment); err != nil {
+		waitFor := initialBackoff
+	retryPublish:
+		for i := 0; i < maxRetries; i++ {
+			select {
+			case <-ctx.Done():
+				return
+			case <-time.After(waitFor):
+				if err := plugin.NodePublishVolume(ctx, assignment); err == nil {
+					break retryPublish
+				}
+			}
+			waitFor = waitFor << 1
+		}
+	}
+}
+
+func (r *volumes) tryRemoveVolume(ctx context.Context, assignment *api.VolumeAssignment) {
```

Comment:
Not going to ask to address in this PR, but we need to add some code to cancel existing attempts at adding a volume if we suddenly need to remove it. So, just adding a TODO comment, like this:

```go
// TODO(ameyag): Cancel existing tryAddVolume when we try to remove a volume
```